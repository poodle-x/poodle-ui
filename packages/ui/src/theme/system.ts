import {
	BordersProps,
	bordersSystem,
	ColorsProps,
	colorsSystem,
	compose,
	createSystem,
	FlexProps,
	flexSystem,
	FontsProps,
	fontsSystem,
	getMode,
	GridProps,
	gridSystem,
	LayoutProps,
	layoutSystem,
	ListProps,
	listSystem,
	PseudosProps,
	pseudosSystem,
	ResponsiveProp,
	ShadowProps,
	shadowSystem,
	SpacingProps,
	spacingSystem,
	SystemConfigList,
} from "@poodle/system";
import * as CSS from "csstype";
import { CSSSystem } from "../styled";
import { StandardThemeConfig } from "./index";

const additionalSystem: SystemConfigList = {
	objectFit: true,
	objectPosition: true,
	transition: true,
	cursor: true,
	pointerEvents: true,
};

export interface AdditionalSystemProps {
	/**
	 * The CSS `object-fit` property
	 */
	objectFit?: ResponsiveProp<CSS.Property.ObjectFit>;
	/**
	 * The CSS `object-fit` property
	 */
	objectPosition?: ResponsiveProp<CSS.Property.ObjectPosition>;
	/**
	 * The CSS `transition` property
	 */
	transition?: ResponsiveProp<CSS.Property.Transition>;
	/**
	 * The CSS `cursor` property
	 */
	cursor?: ResponsiveProp<CSS.Property.Cursor>;
	/**
	 * The CSS `cursor` property
	 */
	pointerEvents?: ResponsiveProp<CSS.Property.PointerEvents>;
}

const allSystem = compose(
	colorsSystem,
	fontsSystem,
	layoutSystem,
	spacingSystem,
	bordersSystem,
	flexSystem,
	pseudosSystem,
	shadowSystem,
	gridSystem,
	listSystem,
	additionalSystem
);

export const system = createSystem(allSystem);

interface CssSystemProps
	extends SpacingProps,
		LayoutProps,
		FontsProps,
		ColorsProps,
		BordersProps,
		ShadowProps,
		FlexProps,
		GridProps,
		ListProps,
		AdditionalSystemProps {}

export interface SystemProps
	extends SpacingProps,
		LayoutProps,
		FontsProps,
		ColorsProps,
		BordersProps,
		ShadowProps,
		FlexProps,
		PseudosProps<CssSystemProps>,
		GridProps,
		ListProps,
		AdditionalSystemProps {}

const additionalProps = [
	"textStyle",
	"variant",
	"themeExtend",
	"theme",
	"sizeStyle",
	"colorStyle",
];

// Cache to avoid unnecessary repeat loop
// @todo maybe move to a helper func
const finalSystem: SystemConfigList = {};

for (const keys in allSystem) {
	const keysPart = keys.split(",");

	const rawConfig = allSystem[keys];

	keysPart.forEach((k) => {
		finalSystem[k] = rawConfig;
	});
}

export function removeSystemProps(props: { [key: string]: any }) {
	let resultProps = {};

	for (const prop in props) {
		let found = false;

		if (finalSystem[prop] || additionalProps.indexOf(prop) >= 0) {
			found = true;
		}

		if (!found && ["sx"].indexOf(prop) === -1) {
			resultProps = {
				...resultProps,
				[prop]: props[prop],
			};
		}
	}

	return resultProps;
}

export function createCSSSystemStandard<Key extends string>(data: {
	key: Key;
	config?: StandardThemeConfig<Key, any>;
	props: { [key: string]: any };
	base?: CSSSystem["base"];
}): CSSSystem {
	const { key, props, config, base } = data;

	const { variant = "", sizeStyle = "", colorStyle = "" } = props;

	const mode = getMode(props);

	return {
		base,
		applies: [
			config?.styles?.[key],
			config?.modes?.[mode]?.styles?.[key],
			config?.variants?.[variant]?.[key],
			config?.modes?.[mode]?.variants?.[variant]?.[key],
			config?.sizes?.[sizeStyle]?.[key],
			config?.modes?.[mode]?.sizes?.[sizeStyle]?.[key],
			config?.colorStyles?.[colorStyle]?.[key],
			config?.modes?.[mode]?.colorStyles?.[colorStyle]?.[key],
		],
		override: config?.overrides?.[key],
	};
}

export * from "@poodle/system";
