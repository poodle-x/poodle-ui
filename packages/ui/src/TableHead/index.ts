import { TableHead, TableHeadProps } from "./TableHead";
import { StandardThemeConfig } from "../theme";

export * from "./TableHead";

export default TableHead;

export type TableHeadStyleKeys = "root";

export interface TableHeadThemeConfig
	extends StandardThemeConfig<TableHeadStyleKeys, TableHeadProps> {}
