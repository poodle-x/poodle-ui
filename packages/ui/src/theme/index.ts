import React from "react";
import {
	ThemeValue,
	SystemThemeConfig,
	BreakpointsThemeConfig,
	SpacingThemeConfig,
	ColorsThemeConfig,
	FontsThemeConfig,
	BordersThemeConfig,
	ShadowThemeConfig,
	LayoutsThemeConfig,
} from "@poodle/system";

import { CSSObjectSystem, Stylesheet } from "../styled";
import { BoxThemeConfig } from "../Box";
import { ButtonThemeConfig } from "../Button";
import { ColumnsThemeConfig } from "../Columns";
import { ColumnThemeConfig } from "../Column";
import { ContainerThemeConfig } from "../Container";
import { TableThemeConfig } from "../Table";
import { TableHeadThemeConfig } from "../TableHead";
import { TableBodyThemeConfig } from "../TableBody";
import { TableRowThemeConfig } from "../TableRow";
import { TableCellThemeConfig } from "../TableCell";
import { VisuallyHiddenThemeConfig } from "../VisuallyHidden";
import { CheckControlThemeConfig } from "../CheckControl";
import { CheckboxThemeConfig } from "../Checkbox";
import { RadioThemeConfig } from "../Radio";
import { InputThemeConfig } from "../Input";
import { InputAdornmentThemeConfig } from "../InputAdornment";
import { IconButtonThemeConfig } from "../IconButton";
import { ModalThemeConfig } from "../Modal";
import { ModalHeaderThemeConfig } from "../ModalHeader";
import { ModalBodyThemeConfig } from "../ModalBody";
import { ModalFooterThemeConfig } from "../ModalFooter";
import { SelectThemeConfig } from "../Select";
import { TabsThemeConfig } from "../Tabs";
import { TabThemeConfig } from "../Tab";
import { TabListThemeConfig } from "../TabList";
import { TabPanelThemeConfig } from "../TabPanel";

export { Stylesheet };

export { useTheme } from "./useTheme";

export type DirectionThemeConfig = "rtl" | "ltr";

export interface GlobalThemeConfig {
	styles?: ThemeValue<CSSObjectSystem>;
}

export type IconType = React.FC<
	React.PropsWithoutRef<React.SVGProps<SVGSVGElement>> &
		React.RefAttributes<SVGSVGElement>
>;

export interface BaseIcons {
	RadioButton?: IconType;
	RadioCheckedButton?: IconType;
	CheckboxButton?: IconType;
	CheckboxCheckedButton?: IconType;
	CheckboxIndeterminateButton?: IconType;
	ArrowDown?: IconType;
}

export interface ThemeConfig extends SystemThemeConfig {
	global?: GlobalThemeConfig;
	direction?: ThemeValue<DirectionThemeConfig>;
	baseIcons?: BaseIcons;
	Box?: BoxThemeConfig;
	Button?: ButtonThemeConfig;
	Columns?: ColumnsThemeConfig;
	Column?: ColumnThemeConfig;
	Container?: ContainerThemeConfig;
	Table?: TableThemeConfig;
	TableHead?: TableHeadThemeConfig;
	TableBody?: TableBodyThemeConfig;
	TableRow?: TableRowThemeConfig;
	TableCell?: TableCellThemeConfig;
	VisuallyHidden?: VisuallyHiddenThemeConfig;
	CheckControl?: CheckControlThemeConfig;
	Checkbox?: CheckboxThemeConfig;
	Radio?: RadioThemeConfig;
	Input?: InputThemeConfig;
	InputAdornment?: InputAdornmentThemeConfig;
	IconButton?: IconButtonThemeConfig;
	Modal?: ModalThemeConfig;
	ModalHeader?: ModalHeaderThemeConfig;
	ModalBody?: ModalBodyThemeConfig;
	ModalFooter?: ModalFooterThemeConfig;
	Select?: SelectThemeConfig;
	Tabs?: TabsThemeConfig;
	Tab?: TabThemeConfig;
	TabList?: TabListThemeConfig;
	TabPanel?: TabPanelThemeConfig;
}

export const defaultGlobal = (
	overrideTheme: ThemeConfig
): GlobalThemeConfig => {
	return {
		styles: overrideTheme.global?.styles,
	};
};

export const defaultBreakpoints = (
	overrideTheme: ThemeConfig
): BreakpointsThemeConfig => {
	return {
		scale: ["sm", "md", "lg", "xl"],
		...overrideTheme.breakpoints,
		sets: {
			sm: 576,
			md: 768,
			lg: 992,
			xl: 1200,
			xxl: 1400,
			...overrideTheme.breakpoints?.sets,
		},
	};
};

export const defaultSpacing = (
	overrideTheme: ThemeConfig
): SpacingThemeConfig => {
	return {
		scale: 4,
		...overrideTheme.spacing,
		sets: {
			s: 4,
			m: 8,
			l: 16,
			...overrideTheme.spacing?.sets,
		},
	};
};

export const defaultFonts = (overrideTheme: ThemeConfig): FontsThemeConfig => {
	return {
		scale: "1rem",
		...overrideTheme.fonts,
		base: {
			fontFamily:
				'-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"',
			fontStyle: "normal",
			fontWeight: "normal",
			lineHeight: "125%",
			fontSize: "1rem",
			...overrideTheme.fonts?.base,
		},
		weights: {
			normal: 400,
			semibold: 600,
			bold: 700,
			...overrideTheme.fonts?.weights,
		},
		sets: {
			input: {
				fontSize: "1rem",
				fontWeight: "normal",
				lineHeight: "125%",
			},
			button: {
				fontSize: "1rem",
				fontWeight: "bold",
				lineHeight: "150%",
			},
			label: {
				fontStyle: "normal",
				fontWeight: "bold",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			helper: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			checkControlLabelMedium: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "1rem",
				lineHeight: "110%",
			},
			checkControlLabelSmall: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			checkControlLabelLarge: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "1.125rem",
				lineHeight: "110%",
			},
			modalTitle: {
				fontStyle: "normal",
				fontWeight: "bold",
				fontSize: "1.5rem",
				lineHeight: "133.33%",
			},
			...overrideTheme.fonts?.sets,
		},
	};
};

export const defaultColor = (overrideTheme: ThemeConfig): ColorsThemeConfig => {
	return {
		modes: {
			...overrideTheme.colors?.modes,
		},
		sets: {
			text: "#282828",
			disabledText: "#A6A6A6",

			primary: {
				light: "#D1EBFF",
				base: "#00478A",
				dark: "#003A70",
				textOnLight: "#00478A",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			negative: {
				base: "#AF1308",
				dark: "#8D1007",
				light: "#FCF3F3",
				textOnLight: "#AF1308",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			positive: {
				base: "#006122",
				dark: "#004D1B",
				light: "#E5FFEE",
				textOnLight: "#006122",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			warn: {
				base: "#FFBB33",
				dark: "#FAAB00",
				light: "#FFF1DB",
				textOnLight: "#282828",
				textOnBase: "#282828",
				textOnDark: "#282828",
			},

			bg: {
				base: "#F8F8F8",
				100: "#E0E0E0",
				textOn100: "#282828",
				200: "#CCCCCC",
				textOn200: "#282828",
			},

			modal: {
				base: "#F8F8F8",
				footer: "#F0F0F0",
			},

			border: {
				base: "#CCCCCC",
			},

			...overrideTheme.colors?.sets,
		},
	};
};

export const defaultBorder = (
	overrideTheme: ThemeConfig
): BordersThemeConfig => {
	return {
		sets: {
			base: {
				style: "solid",
				width: "1px",
				color: "border",
			},
			...overrideTheme?.borders?.sets,
		},
		radius: {
			base: "4px",
			...overrideTheme?.borders?.radius,
		},
	};
};

export const defaultBoxShadow = (
	overrideTheme: ThemeConfig
): ShadowThemeConfig => {
	return {
		boxShadowSets: {
			lv1: "1px 2px 2px -1px rgba(0,0,0,0.4)",
			lv2: "2px 2px 5px 0px rgba(0,0,0,0.4)",
			lv3: "4px 4px 5px 0px rgba(0,0,0,0.4)",
			...overrideTheme?.shadows?.boxShadowSets,
		},
	};
};

export const defaultLayouts = (
	overrideTheme: ThemeConfig
): LayoutsThemeConfig => {
	return {
		zIndices: {
			modal: 2000,
			...overrideTheme?.layouts?.zIndices,
		},
	};
};

export function createTheme(overrideTheme: ThemeConfig = {}): ThemeConfig {
	return {
		direction: "ltr",
		...overrideTheme,
		global: defaultGlobal(overrideTheme),
		spacing: defaultSpacing(overrideTheme),
		breakpoints: defaultBreakpoints(overrideTheme),
		colors: defaultColor(overrideTheme),
		fonts: defaultFonts(overrideTheme),
		borders: defaultBorder(overrideTheme),
		shadows: defaultBoxShadow(overrideTheme),
		layouts: defaultLayouts(overrideTheme),
	};
}

export type StylesConfig<T extends string> = {
	[K in T]: ThemeValue<CSSObjectSystem>;
};

export interface StandardComponentProps {
	variant?: string;
	themeExtend?: ThemeConfig;
	theme?: ThemeConfig;
	sizeStyle?: string;
	colorStyle?: string;
}

export interface StandardThemeConfig<
	T extends string,
	Props extends { [key: string]: any }
> {
	defaultProps?: Partial<Props>;
	styles?: Partial<StylesConfig<T>>;
	variants?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	sizes?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	colorStyles?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	modes?: {
		[key: string]: {
			styles?: Partial<StylesConfig<T>>;
			variants?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
			sizes?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
			colorStyles?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
		};
	};
	overrides?: Partial<StylesConfig<T>>;
	overrideClasses?: Partial<
		{
			[K in T]: boolean | string;
		}
	>;
}

export * from "./system";
