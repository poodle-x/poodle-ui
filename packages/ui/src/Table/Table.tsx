import React from "react";
import Box, { BoxProps } from "../Box";
import useDefaultProps from "../hooks/useDefaultProps/useDefaultProps";
import { getCSSSystemBoxProps } from "../styled/system";
import { StandardComponentProps, ThemeConfig } from "../theme";
import * as styles from "./styles";

const op = {};

export interface TableContextValue {
	withBorder?: boolean;
	withDivider?: boolean;
	withHover?: boolean;
	sizeStyle?: "s" | "m" | "l" | string;
}

export const TableContext = React.createContext<TableContextValue>(op);

export interface TableLevelContextValue {
	level?: "head" | "body" | "footer";
}

export const TableLevelContext = React.createContext<TableLevelContextValue>(
	op
);

export interface LocalTableProps {
	/**
	 * Apply root border.
	 */
	withBorder?: boolean;
	/**
	 * Add divider for each row.
	 */
	withDivider?: boolean;
	/**
	 * Apply hover style on table row.
	 */
	withHover?: boolean;
	/**
	 * Change table size. Built-in options are **"s" | "m" | "l"**.
	 *
	 * Default is **"m"**.
	 */
	sizeStyle?: "s" | "m" | "l" | string;
	/**
	 * Customize inner table props.
	 */
	tableProps?: BoxProps;
	children?: React.ReactNode;
}

export interface TableProps
	extends BoxProps<"table">,
		StandardComponentProps,
		LocalTableProps {}

function getDefaultProps(theme?: ThemeConfig) {
	return theme?.Table?.defaultProps;
}

export const Table: React.ForwardRefExoticComponent<
	React.PropsWithoutRef<TableProps> & React.RefAttributes<HTMLElement>
> = React.forwardRef<HTMLElement, TableProps>((_props, ref) => {
	const { props, isLocalTheme } = useDefaultProps<TableProps>(_props, {
		themeDefaultProps: getDefaultProps,
	});

	const {
		children,
		className,
		tableProps,
		withBorder,
		withDivider,
		withHover,
		sizeStyle = "m",
		...otherProps
	} = props;

	return (
		<Box
			{...otherProps}
			{...getCSSSystemBoxProps({
				isRoot: true,
				isLocalTheme,
				componentProps: props,
				fnCSSSystem: styles.Root,
				baseClassName: ["poodle-table"],
			})}
			ref={ref}
		>
			<Box
				as="table"
				{...getCSSSystemBoxProps({
					isLocalTheme,
					componentProps: props,
					partProps: tableProps,
					fnCSSSystem: styles.Table,
					baseClassName: ["poodle-table__table"],
				})}
			>
				<TableContext.Provider
					value={{
						sizeStyle,
						withBorder,
						withDivider,
						withHover,
					}}
				>
					{children}
				</TableContext.Provider>
			</Box>
		</Box>
	);
});

Table.defaultProps = {
	sizeStyle: "m",
};

Table.displayName = "PoodleTable";
