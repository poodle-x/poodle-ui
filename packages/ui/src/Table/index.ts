import { Table, TableProps } from "./Table";
import { StandardThemeConfig } from "../theme";

export * from "./Table";

export default Table;

export type TableStyleKeys = "root" | "table";

export interface TableThemeConfig
	extends StandardThemeConfig<TableStyleKeys, TableProps> {}
