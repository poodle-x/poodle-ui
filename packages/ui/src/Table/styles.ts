import { createCSSSystemStandard } from "../theme";
import { CSSSystem } from "../styled";
import { TableProps } from "./Table";
import { CSSObjectSystem } from "../styled";
import { TableStyleKeys } from "./";

export function Root(props: TableProps): CSSSystem {
	const { theme, withBorder, withHover } = props;

	const borderStyles: CSSObjectSystem = withBorder
		? {
				borderRadius: "base",
				border: "1px solid",
				borderColor: "border",

				"tr:last-child > td": {
					borderBottom: "0",
				},
		  }
		: {};

	const bgStyles: CSSObjectSystem = withHover
		? {
				"tbody > tr:hover": {
					bg: "bg.100",
					color: "bg.textOn100",
				},

				"tbody > tr[data-selected='true']": {
					bg: "primary.light",
					color: "primary.textOnLight",
				},
		  }
		: {};

	return createCSSSystemStandard<TableStyleKeys>({
		key: "root",
		config: theme?.Table,
		props,
		base: [borderStyles, bgStyles],
	});
}

export function Table(props: TableProps): CSSSystem {
	const { theme } = props;

	return createCSSSystemStandard<TableStyleKeys>({
		key: "table",
		config: theme?.Table,
		props,
		base: [
			{
				width: "100%",
				borderSpacing: 0,
				borderCollapse: "collapse",
				textAlign: "left",
			},
		],
	});
}
