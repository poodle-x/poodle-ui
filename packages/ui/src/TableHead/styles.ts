import { createCSSSystemStandard } from "../theme";
import { CSSSystem } from "../styled";
import { TableHeadProps } from "./TableHead";
import { TableContextValue } from "../Table";
import { TableHeadStyleKeys } from "./";

export function Root(props: TableHeadProps & TableContextValue): CSSSystem {
	const { theme } = props;

	return createCSSSystemStandard<TableHeadStyleKeys>({
		key: "root",
		config: theme?.TableHead,
		props,
		base: [],
	});
}
