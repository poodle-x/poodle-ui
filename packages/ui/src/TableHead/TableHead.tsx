import React from "react";
import Box, { BoxProps } from "../Box";
import useDefaultProps from "../hooks/useDefaultProps/useDefaultProps";
import { getCSSSystemBoxProps } from "../styled/CSSSystem";
import { TableContext, TableLevelContext } from "../Table";
import { StandardComponentProps, ThemeConfig } from "../theme";
import * as styles from "./styles";

export type LocalTableHeadProps = {
	children?: React.ReactNode;
};

export interface TableHeadProps
	extends BoxProps,
		StandardComponentProps,
		LocalTableHeadProps {}

function getDefaultProps(theme?: ThemeConfig) {
	return theme?.TableHead?.defaultProps;
}

export const TableHead: React.ForwardRefExoticComponent<
	React.PropsWithoutRef<TableHeadProps> & React.RefAttributes<HTMLElement>
> = React.forwardRef<HTMLElement, TableHeadProps>((_props, ref) => {
	const { props, isLocalTheme } = useDefaultProps<TableHeadProps>(_props, {
		themeDefaultProps: getDefaultProps,
	});

	const tableContext = React.useContext(TableContext);

	const styleProps = React.useMemo(() => {
		return { ...props, ...tableContext };
	}, [props, tableContext]);

	const { children, className, ...otherProps } = props;

	return (
		<Box
			as="thead"
			{...otherProps}
			{...getCSSSystemBoxProps({
				isRoot: true,
				isLocalTheme,
				componentProps: styleProps,
				fnCSSSystem: styles.Root,
				baseClassName: ["poodle-table-head"],
			})}
			ref={ref}
		>
			<TableLevelContext.Provider value={{ level: "head" }}>
				{children}
			</TableLevelContext.Provider>
		</Box>
	);
});

TableHead.displayName = "PoodleTableHead";
