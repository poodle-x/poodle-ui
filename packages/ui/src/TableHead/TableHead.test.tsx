import React from "react";
import { Table, TableLevelContext } from "../Table";
import { ThemeConfig } from "../theme";
import { render } from "../utils/test";
import TableHead from "./";

describe("<TableHead />", () => {
	it("should render TableHead correctly", () => {
		const { container } = render(
			<table>
				<TableHead>
					<tr>
						<td>Table</td>
					</tr>
				</TableHead>
			</table>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("ref should work", () => {
		const ref = React.createRef<HTMLDivElement>();

		render(
			<table>
				<TableHead ref={ref}>
					<tr>
						<td>Table</td>
					</tr>
				</TableHead>
			</table>
		);
		expect(ref.current).toMatchSnapshot();
	});

	it("should render Box props correctly", () => {
		const { container } = render(
			<table>
				<TableHead id="id" mt="20px">
					<tr>
						<td>Table</td>
					</tr>
				</TableHead>
			</table>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("should send context correctly", () => {
		const { container } = render(
			<table>
				<TableHead>
					<tr>
						<td>
							<TableLevelContext.Consumer>
								{(value) => JSON.stringify(value)}
							</TableLevelContext.Consumer>
						</td>
					</tr>
				</TableHead>
			</table>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	describe("theming", () => {
		const themeTest: ThemeConfig = {
			mode: "mode",
			TableHead: {
				styles: {
					root: {
						mt: 10,
						letterSpacing: 250,
					},
				},
				variants: {
					custom: {
						root: {
							top: 0,
							textIndent: 1,
						},
					},
				},
			},
		};

		it("should extend styles in global", () => {
			const { container } = render(
				<table>
					<TableHead variant="custom">
						<tr>
							<td>Table</td>
						</tr>
					</TableHead>
				</table>,
				{
					theme: themeTest,
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should extend styles in local", () => {
			const { container } = render(
				<table>
					<TableHead variant="custom" themeExtend={themeTest}>
						<tr>
							<td>Table</td>
						</tr>
					</TableHead>
				</table>
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should render with default props correctly", () => {
			const { container } = render(
				<table>
					<TableHead>
						<tr>
							<td>Table</td>
						</tr>
					</TableHead>
				</table>,
				{
					theme: {
						TableHead: {
							defaultProps: {
								id: "id",
								className: "hello",
							},
						},
					},
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should disable classes", () => {
			const { container } = render(
				<table>
					<TableHead>
						<tr>
							<td>Table</td>
						</tr>
					</TableHead>
				</table>,
				{
					theme: {
						TableHead: {
							overrideClasses: {
								root: true,
							},
						},
					},
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should override classes", () => {
			const { container } = render(
				<table>
					<TableHead>
						<tr>
							<td>Table</td>
						</tr>
					</TableHead>
				</table>,
				{
					theme: {
						TableHead: {
							overrideClasses: {
								root: "morning",
							},
						},
					},
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should override styles", () => {
			const { container } = render(
				<table>
					<TableHead>
						<tr>
							<td>Table</td>
						</tr>
					</TableHead>
				</table>,
				{
					theme: {
						TableHead: {
							overrides: {
								root: {
									mt: 50,
								},
							},
						},
					},
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should extend size styles", () => {
			const { container } = render(
				<Table sizeStyle="test">
					<TableHead>
						<tr>
							<td>Table</td>
						</tr>
					</TableHead>
				</Table>,
				{
					theme: {
						TableHead: {
							sizes: {
								test: {
									root: {
										height: "100px",
									},
								},
							},
						},
					},
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should extend size styles with mode", () => {
			const { container } = render(
				<Table sizeStyle="test">
					<TableHead>
						<tr>
							<td>Table</td>
						</tr>
					</TableHead>
				</Table>,
				{
					theme: {
						mode: "test",
						TableHead: {
							sizes: {
								test: {
									root: {
										height: "100px",
									},
								},
							},
							modes: {
								test: {
									sizes: {
										test: {
											root: {
												height: "110px",
											},
										},
									},
								},
							},
						},
					},
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});
	});
});
