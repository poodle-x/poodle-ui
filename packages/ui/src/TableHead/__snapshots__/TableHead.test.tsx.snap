// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableHead /> ref should work 1`] = `
<thead
  class="poodle-box poodle-table-head c0"
>
  <tr>
    <td>
      Table
    </td>
  </tr>
</thead>
`;

exports[`<TableHead /> should render Box props correctly 1`] = `
.c0 {
  margin-top: 20px;
}

<table>
  <thead
    class="poodle-box poodle-table-head c0"
    id="id"
  >
    <tr>
      <td>
        Table
      </td>
    </tr>
  </thead>
</table>
`;

exports[`<TableHead /> should render TableHead correctly 1`] = `
<table>
  <thead
    class="poodle-box poodle-table-head c0"
  >
    <tr>
      <td>
        Table
      </td>
    </tr>
  </thead>
</table>
`;

exports[`<TableHead /> should send context correctly 1`] = `
<table>
  <thead
    class="poodle-box poodle-table-head c0"
  >
    <tr>
      <td>
        {"level":"head"}
      </td>
    </tr>
  </thead>
</table>
`;

exports[`<TableHead /> theming should disable classes 1`] = `
<table>
  <thead
    class="poodle-box poodle-table-head c0"
  >
    <tr>
      <td>
        Table
      </td>
    </tr>
  </thead>
</table>
`;

exports[`<TableHead /> theming should extend size styles 1`] = `
.c1 {
  width: 100%;
  border-spacing: 0;
  border-collapse: collapse;
  text-align: left;
}

.c2 {
  height: 100px;
}

<div
  class="poodle-box poodle-table c0"
>
  <table
    class="poodle-box poodle-table__table c1"
  >
    <thead
      class="poodle-box poodle-table-head c2"
    >
      <tr>
        <td>
          Table
        </td>
      </tr>
    </thead>
  </table>
</div>
`;

exports[`<TableHead /> theming should extend size styles with mode 1`] = `
.c1 {
  width: 100%;
  border-spacing: 0;
  border-collapse: collapse;
  text-align: left;
}

.c2 {
  height: 100px;
  height: 110px;
}

<div
  class="poodle-box poodle-table c0"
>
  <table
    class="poodle-box poodle-table__table c1"
  >
    <thead
      class="poodle-box poodle-table-head c2"
    >
      <tr>
        <td>
          Table
        </td>
      </tr>
    </thead>
  </table>
</div>
`;

exports[`<TableHead /> theming should extend styles in global 1`] = `
.c0 {
  margin-top: 10px;
  letter-spacing: 250px;
  top: 0;
  text-indent: 1px;
}

<table>
  <thead
    class="poodle-box poodle-table-head c0"
  >
    <tr>
      <td>
        Table
      </td>
    </tr>
  </thead>
</table>
`;

exports[`<TableHead /> theming should extend styles in local 1`] = `
.c0 {
  margin-top: 10px;
  letter-spacing: 250px;
  top: 0;
  text-indent: 1px;
}

<table>
  <thead
    class="poodle-box poodle-table-head c0"
  >
    <tr>
      <td>
        Table
      </td>
    </tr>
  </thead>
</table>
`;

exports[`<TableHead /> theming should override classes 1`] = `
<table>
  <thead
    class="poodle-box poodle-table-head c0"
  >
    <tr>
      <td>
        Table
      </td>
    </tr>
  </thead>
</table>
`;

exports[`<TableHead /> theming should override styles 1`] = `
.c0 {
  margin-top: 50px;
}

<table>
  <thead
    class="poodle-box poodle-table-head c0"
  >
    <tr>
      <td>
        Table
      </td>
    </tr>
  </thead>
</table>
`;

exports[`<TableHead /> theming should render with default props correctly 1`] = `
<table>
  <thead
    class="poodle-box poodle-table-head hello c0"
    id="id"
  >
    <tr>
      <td>
        Table
      </td>
    </tr>
  </thead>
</table>
`;
